### PR TITLE
[wip] Fix public protobuf interface - wip

### PR DESCRIPTION
**Pull request by Yangqing**
This is an ongoing fix to the protobuf issue, mainly to address 2 things:

(1) we have random protobuf fixes trying to patch an already flaky system (for example, dual install from anaconda and brew). This diff aims to basically use standard packages as much as possible (protobuf cmake config files, or FindProtoBuf.cmake) and then enforce the build script to explicitly set paths.

(2) We need protobuf to be in the public interface of Caffe2. This PR adds it.

(3) We will most likely need a protobuf diagnostic tool / script. TBD.

Firing a PR so that we can launch build tests.

### CODE REVIEW

File: cmake/public/protobuf.cmake [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  # modern CMake that ships with a FindProtobuf.cmake file that produces
+  # modern targets.
+  message(STATUS "Caffe2: Found protobuf with new-style protobuf targets.")
+elseif(Protobuf_FOUND OR PROTOBUF_FOUND)
+  # If the modern targets are not present, we will generate them for you for
+  # backward compatibility. This is backported from CMake's new FindProtobuf.cmake
+  # content.
+  message(STATUS "Caffe2: Found protobuf with old-style protobuf targets.")
+  if(PROTOBUF_LIBRARY)
+      if(NOT TARGET protobuf::libprotobuf)
```
Comment (pjh5) [inline cmake/public/protobuf.cmake]: Will this question always be true?

File: cmake/public/protobuf.cmake [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      COMMENT "Running C++/Python protocol buffer compiler on ${fil}" VERBATIM )
+  endforeach()
+
+  set_source_files_properties(${${srcs_var}} ${${hdrs_var}} ${${python_var}} PROPERTIES GENERATED TRUE)
+  set(${srcs_var} ${${srcs_var}} PARENT_SCOPE)
+  set(${hdrs_var} ${${hdrs_var}} PARENT_SCOPE)
+  set(${python_var} ${${python_var}} PARENT_SCOPE)
+endfunction()
+
+# After above, we should have the glog::glog target now.
```
Comment (pjh5) [inline cmake/public/protobuf.cmake]: nit: protobuf::libprotobuf

**Comment by Yangqing** [comment]
Discovered issue in centos build that is being fixed at #1965.

File: cmake/public/protobuf.cmake (hunk 2 above)
Comment (Yangqing) [reply cmake/public/protobuf.cmake]: nice catch :)

File: cmake/public/protobuf.cmake (hunk 1 above)
Comment (Yangqing) [reply cmake/public/protobuf.cmake]: I think this should always be true. Let me just remove it, indeed.

**Comment by Yangqing** [comment]
@caffe2bot test this please.

**Comment by Yangqing** [comment]
@caffe2bot  retest this please. Seems that jenkins had a hiccup.

**Comment by Yangqing** [comment]
@caffe2bot retest this please. Jenkins docker update.

**Comment by Yangqing** [comment]
@pjh5 @orionr PTAL? :)

Cc @prigoyal FYI

**Review by pjh5 — APPROVED** [review]
Saw a few comment / style nits but I'm looking forward to getting this merged sooner rather than later

File: cmake/ProtoBuf.cmake [hunk 3]
```diff
@@ -74,72 +48,54 @@ elseif (ANDROID OR IOS)
...
+      "own protobuf under third_party. Note that this behavior may change in "
+      "the future, and you will need to specify -DBUILD_CUSTOM_PROTOBUF=ON "
+      "explicitly.")
   custom_protobuf_find()
-endif()
 
-if(NOT(Protobuf_FOUND OR PROTOBUF_FOUND))
-  message(FATAL_ERROR "Could not find protobuf or compile local version")
+  # TODO(jiayq): enable this in the future, when Jenkins Mac support is
+  # prooerly set up with protobuf installs.
```
Comment (orionr) [inline cmake/ProtoBuf.cmake]: Nit: prooerly -> properly

**Review by orionr — APPROVED** [review]
Yes please!

**Comment by orionr** [comment]
@caffe2bot retest this please

File: cmake/public/protobuf.cmake [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  endif()
+  
+  if(PROTOBUF_PROTOC_EXECUTABLE)
+    add_executable(protobuf::protoc IMPORTED)
+    set_property(TARGET protobuf::protoc PROPERTY
+        IMPORTED_LOCATION ${PROTOBUF_PROTOC_EXECUTABLE})
+  endif()
+endif()
+
+# After above, we should have the glog::glog target now.
```
Comment (pietern) [inline cmake/public/protobuf.cmake]: This says glog but should be protobuf

File: .jenkins/build.sh [hunk 5]
```diff
@@ -91,6 +91,16 @@ if [ "$(uname)" == "Linux" ]; then
...
+# Currently, on Jenkins mac os, we will use custom protobuf. Mac OS
+# contbuild at the moment is minimal dependency - it doesn't use glog
+# or gflags either.
+if [ "$(uname)" == "Darwin" ]; then
+  CMAKE_ARGS+=("-DBUILD_CUSTOM_PROTOBUF=ON")
+fi
+
+
+
+
```
Comment (pietern) [inline .jenkins/build.sh]: Is this still debugging stuff? Extra newlines can go.

File: cmake/Caffe2Config.cmake.in [hunk 6]
```diff
@@ -38,6 +38,26 @@ if (@USE_GLOG@)
...
 endif()
 
+# Protobuf
+include("${CMAKE_CURRENT_LIST_DIR}/public/protobuf.cmake")
+if (NOT TARGET protobuf::libprotobuf)
+  message(FATAL_ERROR
+      "Your installed Caffe2 version uses protobuf but the protobuf library "
+      "cannot be found. Did you accidentally remove it, or have you set "
+      "the right CMAKE_PREFIX_PATH? If you do not have protobuf, you will "
+      "need to install glog and set the library path accordingly.")
```
Comment (pietern) [inline cmake/Caffe2Config.cmake.in]: glog -> protobuf

File: cmake/ProtoBuf.cmake [hunk 7]
```diff
@@ -1,68 +1,42 @@
...
+  # result, for older versions, we will manually add alias.
+  if (NOT TARGET protobuf::libprotobuf)
+    add_library(protobuf::libprotobuf ALIAS libprotobuf)
+    add_library(protobuf::libprotobuf-lite ALIAS libprotobuf-lite)
+    add_executable(protobuf::protoc ALIAS protoc)
   endif()
-elseif (ANDROID OR IOS)
+endmacro()
+
+message(STATUS "debug: BUILD_CUSTOM_PROTOBUF is " ${BUILD_CUSTOM_PROTOBUF})
```
Comment (pietern) [inline cmake/ProtoBuf.cmake]: Should go before merging

File: cmake/public/protobuf.cmake [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  # Protobuf config file installed (like on Windows), or you are using a
+  # modern CMake that ships with a FindProtobuf.cmake file that produces
+  # modern targets.
+  message(STATUS "Caffe2: Found protobuf with new-style protobuf targets.")
+elseif(Protobuf_FOUND OR PROTOBUF_FOUND)
+  # If the modern targets are not present, we will generate them for you for
+  # backward compatibility. This is backported from CMake's new FindProtobuf.cmake
+  # content.
+  message(STATUS "Caffe2: Found protobuf with old-style protobuf targets.")
+  if(PROTOBUF_LIBRARY)
```
Comment (pietern) [inline cmake/public/protobuf.cmake]: Trailing whitespace

**Review by pietern — COMMENTED** [review]
Nice work! Mostly typo's in this review.

**Comment by Yangqing** [comment]
Thanks folks - will merge when the test passes.

**Comment by Yangqing** [comment]
Confirmed that test failure is due to docker image building; new docker images are being built here: https://ci.pytorch.org/jenkins/job/caffe2-docker-trigger/65/. Landing.

**Pull request merged**